include/ginkgo/core/matrix/scaled_permutation.hpp

```diff
@@ -0,0 +1,203 @@
+#include <ginkgo/core/matrix/permutation.hpp>
+
+
+namespace gko {
+namespace matrix {
+
+
+/**
+ * ScaledPermutation is a matrix combining a permutation with scaling factors.
+ * It is a combination of Diagonal and Permutation, and can be read as
+ * $SP = P \pdot S$, i.e. the scaling gets applied before the permutation.
+ *
+ * @tparam IndexType  index type of permutation indices
+ * @tparam ValueType  value type of the scaling factors
+ *
+ * @ingroup permutation
+ * @ingroup mat_formats
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class ScaledPermutation
+    : public EnableLinOp<ScaledPermutation<ValueType, IndexType>>,
+      public WritableToMatrixData<ValueType, IndexType> {
+    friend class EnablePolymorphicObject<ScaledPermutation, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+
+    /**
+     * Returns a pointer to the scaling factors.
+     *
+     * @return the pointer to the scaling factors.
+     */
+    value_type* get_scale() noexcept { return scale_.get_data(); }
+
+    /**
+     * @copydoc get_scale()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const value_type* get_const_scale() const noexcept
+    {
+        return scale_.get_const_data();
+    }
+
+    /**
+     * Returns a pointer to the permutation indices.
+     *
+     * @return the pointer to the permutation indices.
+     */
+    index_type* get_permutation() noexcept { return permutation_.get_data(); }
+
+    /**
+     * @copydoc get_permutation()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const index_type* get_const_permutation() const noexcept
+    {
+        return permutation_.get_const_data();
+    }
+
+    /**
+     * Returns the inverse scaled permutation.
+     *
+     * @return a newly created ScaledPermutation object storing the inverse
+     *         permutation and scaling factors of this ScalingPermutation.
+     */
+    std::unique_ptr<ScaledPermutation> invert() const;
+
+    /**
+     * Combines this scaled permutation with another scaled permutation via
+     * composition. This means `result = other * this` from the matrix
+     * perspective, which is equivalent to first scaling and permuting by `this`
+     * and then by `other`.
+     *
+     * @param other  the other permutation
+     * @return the combined permutation
+     */
+    std::unique_ptr<ScaledPermutation> combine(
+        ptr_param<const ScaledPermutation> other) const;
+
+    void write(gko::matrix_data<value_type, index_type>& data) const override;
+
+    /**
+     * Creates an uninitialized ScaledPermutation matrix.
```

Comment:
I actually use it in a few places, it's mostly for convenience and consistency - it saves you a few lines to set up the arrays, and we have the same kind of constructors for all matrix types. But it's a valid question. You could work with just a single constructor that takes arrays, but that would make things like initializing row_ptrs to avoid illegal memory accesses on newly created objects harder.